### PR TITLE
vcs: fix resources deallocation of simulation

### PR DESCRIPTION
**Pull request by xiaokamikami**
When running with --workload-lilst, simv_init is reinitialized multiple times in a single startup, and the pointer from the last simv_init request needs to be destroyed before doing so. This commit fixed an issue with incomplete memory deallocation when exiting the program, and release the list file correctly when using the workload list

### CODE REVIEW

**Comment by klin02** [comment]
Wrap these finish functions into simv_finish, corresponding to simv_init.
By the way, fclose workload list can be also fixed in this PR？The title may be changed to vcs : free resources of simulation?

**Comment by xiaokamikami** [comment]
> Wrap these finish functions into simv_finish, corresponding to simv_init. By the way, fclose workload list can be also fixed in this PR？The title may be changed to vcs : free resources of simulation?

Look at the static FILE* variable you used earlier. It's initialized only once, and you don't open a new file handle every time you call the function, so you don't need to change it

**Comment by klin02** [comment]
> > Wrap these finish functions into simv_finish, corresponding to simv_init. By the way, fclose workload list can be also fixed in this PR？The title may be changed to vcs : free resources of simulation?
> 
> Look at the static FILE* variable you used earlier. It's initialized only once, and you don't open a new file handle every time you call the function, so you don't need to change it

For safety we still need to free the file handle at the end of the program. You don't need to free it in simv_finish（）, but in switch_workload（）

File: src/test/csrc/vcs/vcs_main.cpp [hunk 1]
```diff
@@ -177,6 +177,13 @@ void difftest_deferred_result(uint8_t result) {
 }
 #endif // CONFIG_DIFFTEST_DEFERRED_RESULT
 
+void simv_finish() {
+  difftest_finish();
+  flash_finish();
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:182]: simMemory should be also free, corresponding to `init_ram`, you can check the logic in emu.cpp.
Also, the logic should be the same as simv_init, as a general function. use enable_difftest as if condition.

Comment (xiaokamikami) [reply src/test/csrc/vcs/vcs_main.cpp]: ok

**Comment by klin02** [comment]
https://github.com/OpenXiangShan/difftest/blob/bb991c9ee083c4e229c78cf1d4f2819f574c3852/src/test/csrc/verilator/emu.cpp#L503

File: src/test/csrc/common/ram.cpp [hunk 2]
```diff
@@ -36,6 +36,12 @@ void init_ram(const char *image, uint64_t ram_size) {
   simMemory = new MmapMemory(image, ram_size);
 }
 
+void ram_finish() {
+#ifdef WITH_DRAMSIM3
```
Comment (klin02) [inline src/test/csrc/common/ram.cpp:40]: It seems we will always call init_goldenmem in simv_init when enable_difftest?

File: src/test/csrc/common/ram.cpp [hunk 3]
```diff
@@ -36,6 +36,12 @@ void init_ram(const char *image, uint64_t ram_size) {
...
+  dramsim3_finish();
+#endif
```
Comment (klin02) [inline src/test/csrc/common/ram.cpp:42]: No need to declare another func? Like emu is OK

File: src/test/csrc/vcs/vcs_main.cpp [hunk 4]
```diff
@@ -177,12 +179,20 @@ void difftest_deferred_result(uint8_t result) {
...
+  }
+}
+
 static uint8_t simv_result = SIMV_RUN;
 #ifdef CONFIG_DIFFTEST_DEFERRED_RESULT
 extern "C" void simv_nstep(uint8_t step) {
-  if (simv_result == SIMV_FAIL)
-    return;
-  if (difftest == NULL)
+  if (simv_result == SIMV_FAIL || difftest == NULL)
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:195]: Is it necessary for changed?

Comment (xiaokamikami) [reply src/test/csrc/vcs/vcs_main.cpp:195]: It more concise

Comment (klin02) [reply src/test/csrc/vcs/vcs_main.cpp:198]: This change seems irrelevant to this PR. Such change may bring extra troubles for git. blame

File: src/test/csrc/vcs/vcs_main.cpp [hunk 5]
```diff
@@ -177,12 +179,23 @@ void difftest_deferred_result(uint8_t result) {
...
 #endif // CONFIG_DIFFTEST_DEFERRED_RESULT
 
+void simv_finish() {
+  common_finish();
+  flash_finish();
+  if (enable_difftest) {
+    goldenmem_finish();
+    difftest_finish();
+  }
+#ifdef WITH_DRAMSIM3
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:189]: dram_finish will call automatically with destructor of simMemory?
See https://github.com/OpenXiangShan/difftest/blob/bb991c9ee083c4e229c78cf1d4f2819f574c3852/src/test/csrc/common/ram.cpp#L298C1-L298C11

**Comment by klin02** [comment]
> https://github.com/OpenXiangShan/difftest/blob/bb991c9ee083c4e229c78cf1d4f2819f574c3852/src/test/csrc/verilator/emu.cpp#L503

Click the link to see detail code. After delete, we should set it to nullptr.

**Comment by klin02** [comment]
Try it in palladium with large size workload-list to test memory deallocation. Then I have no more problem.

**Comment by xiaokamikami** [comment]
> Try it in palladium with large size workload-list to test memory deallocation. Then I have no more problem.

ok, I tested a set of workload-lists on palladium with runtime memory usage ranging from 6G to 16G without leaking memory

**Review by klin02 — APPROVED** [review]

**Pull request merged**